tests/circuit_compilation/test_compilation_utils.py

```diff
@@ -0,0 +1,162 @@
+# Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
+# All rights reserved.
+#
+# SPDX-License-Identifier: MIT
+#
+# Licensed under the MIT License
+
+"""Tests for utility functions in compilation_utils."""
+
+import pytest
+from qiskit import QuantumCircuit
+
+from mqt.qecc.circuit_compilation.compilation_utils import (
+    count_code_switches,
+    insert_switch_placeholders,
+)
+
+# ==============================================================================
+# Tests for count_code_switches
+# ==============================================================================
+
+
+def test_count_switches_single_qubit_simple():
+    """Test simple H -> T transition on a single qubit."""
+    qc = QuantumCircuit(1)
+    qc.h(0)  # Code A
+    qc.t(0)  # Code B -> Expect 1 switch
+
+    count, final_codes = count_code_switches(qc)
+
+    assert count == 1
+    assert final_codes[0] == "B"
+
+
+def test_count_switches_no_switch_needed():
+    """Test a sequence that stays within one code."""
+    qc = QuantumCircuit(2)
+    qc.h(0)
+    qc.cx(0, 1)  # H and CX are both supported by Code A
+
+    count, final_codes = count_code_switches(qc)
+
+    assert count == 0
+    assert final_codes[0] == "A"
+    assert final_codes[1] == "A"
+
+
+def test_count_switches_cnot_synchronization():
+    """Test that CNOT forces qubits into the same code."""
+    qc = QuantumCircuit(2)
+    # Q0 starts in Code A (H gate)
+    qc.h(0)
+    # Q1 starts in Code B (T gate)
+    qc.t(1)
+
+    # CNOT requires them to be in the SAME code.
+    # The implementation defaults to switching to A if possible, or B.
+    # Here, one of them MUST switch.
+    qc.cx(0, 1)
+
+    count, _ = count_code_switches(qc)
+    assert count == 1
+
+
+def test_count_switches_invalid_gate():
+    """Test that an unsupported gate raises ValueError."""
+    qc = QuantumCircuit(1)
+    qc.x(0)  # 'x' is not in the defined sets for Code A or B in the utils
```

Comment:
In the context of code switching somewhat of a strange test, since both 2D and 3D color code support an X gate. I know why you do it like this but just a bit odd when I stumbled over it.